src/librustc_typeck/check/coercion.rs

```diff
@@ -355,6 +355,20 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
 
         // Now apply the autoref. We have to extract the region out of
         // the final ref type we got.
+        if ty == a && mt_a.mutbl == hir::MutImmutable && autoderefs == 1 {
```

Comment:
Actually I think the reasoning was a bit less subtle. That is, basically, if the origin is `&T`, we know the origin value is copy, so there is never a need to borrow so as to prevent a move. It so happens that the destination also must be `&T`, but that's neither here nor there.